pkg/control/control.go

```diff
@@ -142,7 +144,74 @@ ROUND:
 	c.tearDownDB()
 }
 
-func (c *Controller) syncExec(f func(i int)) {
+// RunWithServiceQualityProf runs the controller with profiling service quality
+func (c *Controller) RunWithServiceQualityProf() {
+	c.setUpDB()
+	c.setUpClient()
+
+ENTRY:
+	for _, g := range c.nemesisGenerators {
+		for round := 1; round <= c.cfg.RunRound; round++ {
+			log.Printf("%s round %d start ...", g.Name(), round)
+
+			ctx, cancel := context.WithTimeout(c.ctx, c.cfg.RunTime)
+			historyFile := fmt.Sprintf("%s.%s.%d", c.cfg.History, g.Name(), round)
+			recorder, err := history.NewRecorder(historyFile)
+			if err != nil {
+				log.Fatalf("prepare history failed %v", err)
+			}
+
+			if err := c.dumpState(ctx, recorder); err != nil {
+				log.Fatalf("dump state failed %v", err)
+			}
+
+			// requestCount for the round, shared by all clients.
+			requestCount := int64(c.cfg.RequestCount)
+			log.Printf("total request count %d", requestCount)
+
+			n := len(c.cfg.ClientNodes)
+			var clientWg sync.WaitGroup
+			clientWg.Add(n)
+			for i := 0; i < n; i++ {
+				go func(i int) {
+					defer clientWg.Done()
+					c.onClientLoop(ctx, i, &requestCount, recorder)
+				}(i)
+			}
+
+			var nemesisWg sync.WaitGroup
+			nemesisWg.Add(1)
+			go func() {
+				defer nemesisWg.Done()
+				time.Sleep(time.Second * 10)
+				c.dispatchNemesisWithRecord(ctx, g, recorder)
+			}()
+
+			clientWg.Wait()
+			log.Printf("%s round %d client requests done", g.Name(), round)
+			cancel()
+			nemesisWg.Wait()
```

Comment:
Why didn't nemesis done before `clinetWg`?

Reply:
The sequence doesn't matter here. I just ensure that verify is after clients and nemesis all finished.